In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv("Air Pollution Deaths.csv")
df

C:\Users\tarun\anaconda3\lib\site-packages\pandas\core\computation\expressions.py:21: UserWarning: Pandas requires version '2.8.4' or newer of 'numexpr' (version '2.7.3' currently installed).
  from pandas.core.computation.check import NUMEXPR_INSTALLED
C:\Users\tarun\anaconda3\lib\site-packages\pandas\core\arrays\masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.2' currently installed).
  from pandas.core import (


Entity Code  Year  Air pollution (total) (deaths per 100,000)  \
0    Afghanistan  AFG  2021                                  191.801119   
1    Afghanistan  AFG  2022                                  186.668949   
2    Afghanistan  AFG  2023                                  183.941387   
3        Albania  ALB  2021                                   42.299773   
4        Albania  ALB  2022                                   41.018788   
..           ...  ...   ...                                         ...   
688       Zambia  ZMB  2022                                  100.566474   
689       Zambia  ZMB  2023                                   98.869963   
690     Zimbabwe  ZWE  2021                                  132.752553   
691     Zimbabwe  ZWE  2022                                  128.692138   
692     Zimbabwe  ZWE  2023                                  125.028843   

     Indoor air pollution (deaths per 100,000)  \
0                                   141.354238   
1                                   138.663543   
2                                   134.993753   
3                                    19.163116   
4                                    18.846585   
..                                         ...   
688                                  76.537725   
689                                  74.844720   
690                                 104.340506   
691                                 100.392287   
692                                  96.235996   

     Outdoor particulate matter (deaths per 100,000)  \
0                                          47.334537   
1                                          44.826196   
2                                          45.737662   
3                                          21.786089   
4                                          20.810664   
..                                               ...   
688                                        21.534921   
689                                        21.537253   
690                                        25.715415   
691                                        25.643570   
692                                        26.166182   

     Outdoor ozone pollution (deaths per 100,000)  
0                                        5.746082  
1                                        5.779466  
2                                        5.810625  
3                                        1.806165  
4                                        1.805548  
..                                            ...  
688                                      4.043370  
689                                      4.015376  
690                                      4.200907  
691                                      4.117173  
692                                      4.052495  

[693 rows x 7 columns]

In [2]:
import pandas as pd

df.rename(columns={'Entity': 'countries'}, inplace=True)
df


countries Code  Year  Air pollution (total) (deaths per 100,000)  \
0    Afghanistan  AFG  2021                                  191.801119   
1    Afghanistan  AFG  2022                                  186.668949   
2    Afghanistan  AFG  2023                                  183.941387   
3        Albania  ALB  2021                                   42.299773   
4        Albania  ALB  2022                                   41.018788   
..           ...  ...   ...                                         ...   
688       Zambia  ZMB  2022                                  100.566474   
689       Zambia  ZMB  2023                                   98.869963   
690     Zimbabwe  ZWE  2021                                  132.752553   
691     Zimbabwe  ZWE  2022                                  128.692138   
692     Zimbabwe  ZWE  2023                                  125.028843   

     Indoor air pollution (deaths per 100,000)  \
0                                   141.354238   
1                                   138.663543   
2                                   134.993753   
3                                    19.163116   
4                                    18.846585   
..                                         ...   
688                                  76.537725   
689                                  74.844720   
690                                 104.340506   
691                                 100.392287   
692                                  96.235996   

     Outdoor particulate matter (deaths per 100,000)  \
0                                          47.334537   
1                                          44.826196   
2                                          45.737662   
3                                          21.786089   
4                                          20.810664   
..                                               ...   
688                                        21.534921   
689                                        21.537253   
690                                        25.715415   
691                                        25.643570   
692                                        26.166182   

     Outdoor ozone pollution (deaths per 100,000)  
0                                        5.746082  
1                                        5.779466  
2                                        5.810625  
3                                        1.806165  
4                                        1.805548  
..                                            ...  
688                                      4.043370  
689                                      4.015376  
690                                      4.200907  
691                                      4.117173  
692                                      4.052495  

[693 rows x 7 columns]

In [3]:
import pandas as pd

unique_years = df['Year'].unique().tolist()

print(unique_years)


[2021, 2022, 2023]


In [4]:
import pandas as pd

unique_countries = df['countries'].unique().tolist()

print(unique_countries)

['Afghanistan', 'Albania', 'Algeria', 'American Samoa', 'Andean Latin America', 'Andorra', 'Angola', 'Antigua and Barbuda', 'Argentina', 'Armenia', 'Australasia', 'Australia', 'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain', 'Bangladesh', 'Barbados', 'Belarus', 'Belgium', 'Belize', 'Benin', 'Bermuda', 'Bhutan', 'Bolivia', 'Bosnia and Herzegovina', 'Botswana', 'Brazil', 'Brunei', 'Bulgaria', 'Burkina Faso', 'Burundi', 'Cambodia', 'Cameroon', 'Canada', 'Cape Verde', 'Caribbean', 'Central African Republic', 'Central Asia', 'Central Europe', 'Central Europe, Eastern Europe, and Central Asia', 'Central Latin America', 'Central Sub-Saharan Africa', 'Chad', 'Chile', 'China', 'Colombia', 'Comoros', 'Congo', 'Costa Rica', "Cote d'Ivoire", 'Croatia', 'Cuba', 'Cyprus', 'Czech Republic', 'Democratic Republic of Congo', 'Denmark', 'Djibouti', 'Dominica', 'Dominican Republic', 'East Asia', 'Eastern Europe', 'Eastern Sub-Saharan Africa', 'Ecuador', 'Egypt', 'El Salvador', 'England', 'Equatorial Guinea'

In [5]:
import pandas as pd
import datetime


# Function to calculate Unix timestamp for start of the year
def start_of_year_unix(year):
    start_of_year = datetime.datetime(year, 1, 1, 0, 0, 0)
    return int(start_of_year.timestamp())

# Function to calculate Unix timestamp for end of the year (1 second later than start time)
def end_of_year_unix(year):
    start_of_year = datetime.datetime(year, 1, 1, 0, 0, 0)
    end_of_year = start_of_year + datetime.timedelta(seconds=400)
    return int(end_of_year.timestamp())

# Apply the functions to create new columns
df['starttime'] = df['Year'].apply(start_of_year_unix)
df['endtime'] = df['Year'].apply(end_of_year_unix)

df


countries Code  Year  Air pollution (total) (deaths per 100,000)  \
0    Afghanistan  AFG  2021                                  191.801119   
1    Afghanistan  AFG  2022                                  186.668949   
2    Afghanistan  AFG  2023                                  183.941387   
3        Albania  ALB  2021                                   42.299773   
4        Albania  ALB  2022                                   41.018788   
..           ...  ...   ...                                         ...   
688       Zambia  ZMB  2022                                  100.566474   
689       Zambia  ZMB  2023                                   98.869963   
690     Zimbabwe  ZWE  2021                                  132.752553   
691     Zimbabwe  ZWE  2022                                  128.692138   
692     Zimbabwe  ZWE  2023                                  125.028843   

     Indoor air pollution (deaths per 100,000)  \
0                                   141.354238   
1                                   138.663543   
2                                   134.993753   
3                                    19.163116   
4                                    18.846585   
..                                         ...   
688                                  76.537725   
689                                  74.844720   
690                                 104.340506   
691                                 100.392287   
692                                  96.235996   

     Outdoor particulate matter (deaths per 100,000)  \
0                                          47.334537   
1                                          44.826196   
2                                          45.737662   
3                                          21.786089   
4                                          20.810664   
..                                               ...   
688                                        21.534921   
689                                        21.537253   
690                                        25.715415   
691                                        25.643570   
692                                        26.166182   

     Outdoor ozone pollution (deaths per 100,000)   starttime     endtime  
0                                        5.746082  1609477200  1609477600  
1                                        5.779466  1641013200  1641013600  
2                                        5.810625  1672549200  1672549600  
3                                        1.806165  1609477200  1609477600  
4                                        1.805548  1641013200  1641013600  
..                                            ...         ...         ...  
688                                      4.043370  1641013200  1641013600  
689                                      4.015376  1672549200  1672549600  
690                                      4.200907  1609477200  1609477600  
691                                      4.117173  1641013200  1641013600  
692                                      4.052495  1672549200  1672549600  

[693 rows x 9 columns]

In [6]:
import pandas as pd
from geopy.geocoders import Nominatim

# Initialize geocoder
geolocator = Nominatim(user_agent="my_geocoder")

# Function to get latitude and longitude for a country
def get_lat_long(country):
    location = geolocator.geocode(country)
    if location:
        return location.latitude, location.longitude
    else:
        return None, None

# Apply the function to the 'Country' column
df['Latitude'], df['Longitude'] = zip(*df['countries'].apply(get_lat_long))

# Display the DataFrame with latitude and longitude columns
df


countries Code  Year  Air pollution (total) (deaths per 100,000)  \
0    Afghanistan  AFG  2021                                  191.801119   
1    Afghanistan  AFG  2022                                  186.668949   
2    Afghanistan  AFG  2023                                  183.941387   
3        Albania  ALB  2021                                   42.299773   
4        Albania  ALB  2022                                   41.018788   
..           ...  ...   ...                                         ...   
688       Zambia  ZMB  2022                                  100.566474   
689       Zambia  ZMB  2023                                   98.869963   
690     Zimbabwe  ZWE  2021                                  132.752553   
691     Zimbabwe  ZWE  2022                                  128.692138   
692     Zimbabwe  ZWE  2023                                  125.028843   

     Indoor air pollution (deaths per 100,000)  \
0                                   141.354238   
1                                   138.663543   
2                                   134.993753   
3                                    19.163116   
4                                    18.846585   
..                                         ...   
688                                  76.537725   
689                                  74.844720   
690                                 104.340506   
691                                 100.392287   
692                                  96.235996   

     Outdoor particulate matter (deaths per 100,000)  \
0                                          47.334537   
1                                          44.826196   
2                                          45.737662   
3                                          21.786089   
4                                          20.810664   
..                                               ...   
688                                        21.534921   
689                                        21.537253   
690                                        25.715415   
691                                        25.643570   
692                                        26.166182   

     Outdoor ozone pollution (deaths per 100,000)   starttime     endtime  \
0                                        5.746082  1609477200  1609477600   
1                                        5.779466  1641013200  1641013600   
2                                        5.810625  1672549200  1672549600   
3                                        1.806165  1609477200  1609477600   
4                                        1.805548  1641013200  1641013600   
..                                            ...         ...         ...   
688                                      4.043370  1641013200  1641013600   
689                                      4.015376  1672549200  1672549600   
690                                      4.200907  1609477200  1609477600   
691                                      4.117173  1641013200  1641013600   
692                                      4.052495  1672549200  1672549600   

      Latitude  Longitude  
0    33.768006  66.238514  
1    33.768006  66.238514  
2    33.768006  66.238514  
3     1.231526 -75.892043  
4     1.231526 -75.892043  
..         ...        ...  
688 -14.518912  27.558988  
689 -14.518912  27.558988  
690 -18.455496  29.746841  
691 -18.455496  29.746841  
692 -18.455496  29.746841  

[693 rows x 11 columns]

In [7]:
df = df.dropna()
df = df.reset_index(drop=True)
df

countries Code  Year  Air pollution (total) (deaths per 100,000)  \
0    Afghanistan  AFG  2021                                  191.801119   
1    Afghanistan  AFG  2022                                  186.668949   
2    Afghanistan  AFG  2023                                  183.941387   
3        Albania  ALB  2021                                   42.299773   
4        Albania  ALB  2022                                   41.018788   
..           ...  ...   ...                                         ...   
583       Zambia  ZMB  2022                                  100.566474   
584       Zambia  ZMB  2023                                   98.869963   
585     Zimbabwe  ZWE  2021                                  132.752553   
586     Zimbabwe  ZWE  2022                                  128.692138   
587     Zimbabwe  ZWE  2023                                  125.028843   

     Indoor air pollution (deaths per 100,000)  \
0                                   141.354238   
1                                   138.663543   
2                                   134.993753   
3                                    19.163116   
4                                    18.846585   
..                                         ...   
583                                  76.537725   
584                                  74.844720   
585                                 104.340506   
586                                 100.392287   
587                                  96.235996   

     Outdoor particulate matter (deaths per 100,000)  \
0                                          47.334537   
1                                          44.826196   
2                                          45.737662   
3                                          21.786089   
4                                          20.810664   
..                                               ...   
583                                        21.534921   
584                                        21.537253   
585                                        25.715415   
586                                        25.643570   
587                                        26.166182   

     Outdoor ozone pollution (deaths per 100,000)   starttime     endtime  \
0                                        5.746082  1609477200  1609477600   
1                                        5.779466  1641013200  1641013600   
2                                        5.810625  1672549200  1672549600   
3                                        1.806165  1609477200  1609477600   
4                                        1.805548  1641013200  1641013600   
..                                            ...         ...         ...   
583                                      4.043370  1641013200  1641013600   
584                                      4.015376  1672549200  1672549600   
585                                      4.200907  1609477200  1609477600   
586                                      4.117173  1641013200  1641013600   
587                                      4.052495  1672549200  1672549600   

      Latitude  Longitude  
0    33.768006  66.238514  
1    33.768006  66.238514  
2    33.768006  66.238514  
3     1.231526 -75.892043  
4     1.231526 -75.892043  
..         ...        ...  
583 -14.518912  27.558988  
584 -14.518912  27.558988  
585 -18.455496  29.746841  
586 -18.455496  29.746841  
587 -18.455496  29.746841  

[588 rows x 11 columns]

In [8]:
import pandas as pd
import requests

# Function to call the API and extract required data
def fetch_air_pollution_data(row):
    url = "http://api.openweathermap.org/data/2.5/air_pollution/history"
    params = {
        "lat": row['Latitude'],
        "lon": row['Longitude'],
        "start": row['starttime'],
        "end": row['endtime'],
        "appid": "936c8c5a35fb206b715cdfcd87cba32b"
    }
    response = requests.get(url, params=params)
    if response.status_code == 200:
        api_data = response.json()
        if 'list' in api_data and len(api_data['list']) > 0:
            air_quality_data = api_data['list'][0]
            components = air_quality_data['components']
            main = air_quality_data['main']
            # Create new columns
            row['co'] = components.get('co', None)
            row['no'] = components.get('no', None)
            row['no2'] = components.get('no2', None)
            row['o3'] = components.get('o3', None)
            row['so2'] = components.get('so2', None)
            row['pm2_5'] = components.get('pm2_5', None)
            row['pm10'] = components.get('pm10', None)
            row['nh3'] = components.get('nh3', None)
            row['aqi'] = main.get('aqi', None)
    return row

# Apply the function to each row
df = df.apply(fetch_air_pollution_data, axis=1)

# Display the updated DataFrame
df


countries Code  Year  Air pollution (total) (deaths per 100,000)  \
0    Afghanistan  AFG  2021                                  191.801119   
1    Afghanistan  AFG  2022                                  186.668949   
2    Afghanistan  AFG  2023                                  183.941387   
3        Albania  ALB  2021                                   42.299773   
4        Albania  ALB  2022                                   41.018788   
..           ...  ...   ...                                         ...   
583       Zambia  ZMB  2022                                  100.566474   
584       Zambia  ZMB  2023                                   98.869963   
585     Zimbabwe  ZWE  2021                                  132.752553   
586     Zimbabwe  ZWE  2022                                  128.692138   
587     Zimbabwe  ZWE  2023                                  125.028843   

     Indoor air pollution (deaths per 100,000)  \
0                                   141.354238   
1                                   138.663543   
2                                   134.993753   
3                                    19.163116   
4                                    18.846585   
..                                         ...   
583                                  76.537725   
584                                  74.844720   
585                                 104.340506   
586                                 100.392287   
587                                  96.235996   

     Outdoor particulate matter (deaths per 100,000)  \
0                                          47.334537   
1                                          44.826196   
2                                          45.737662   
3                                          21.786089   
4                                          20.810664   
..                                               ...   
583                                        21.534921   
584                                        21.537253   
585                                        25.715415   
586                                        25.643570   
587                                        26.166182   

     Outdoor ozone pollution (deaths per 100,000)   starttime     endtime  \
0                                        5.746082  1609477200  1609477600   
1                                        5.779466  1641013200  1641013600   
2                                        5.810625  1672549200  1672549600   
3                                        1.806165  1609477200  1609477600   
4                                        1.805548  1641013200  1641013600   
..                                            ...         ...         ...   
583                                      4.043370  1641013200  1641013600   
584                                      4.015376  1672549200  1672549600   
585                                      4.200907  1609477200  1609477600   
586                                      4.117173  1641013200  1641013600   
587                                      4.052495  1672549200  1672549600   

      Latitude  Longitude      co    no   no2     o3   so2  pm2_5  pm10   nh3  \
0    33.768006  66.238514  183.58  0.15  0.35  59.37  0.12   2.14  2.56  1.03   
1    33.768006  66.238514  260.35  0.00  0.17  85.12  0.23   3.67  4.16  0.19   
2    33.768006  66.238514  190.26  0.15  0.37  67.95  0.09   2.22  2.58  1.01   
3     1.231526 -75.892043  140.19  0.01  0.24   0.67  0.06   0.54  0.68  0.21   
4     1.231526 -75.892043  146.87  0.02  1.32   2.06  0.16   2.36  2.77  1.65   
..         ...        ...     ...   ...   ...    ...   ...    ...   ...   ...   
583 -14.518912  27.558988  253.68  0.06  0.66   6.79  0.06   2.51  3.34  0.17   
584 -14.518912  27.558988  283.72  0.00  0.43   6.88  0.26   1.15  1.36  0.09   
585 -18.455496  29.746841  380.52  0.02  0.33   2.84  0.43   0.91  1.18  0.16   
586 -18.455496  29.746841  220.30  0.02  0.49  16.99  0.47   1.49  2.02  0.27   
587 -18.455496  

In [9]:
pip install geopy


Note: you may need to restart the kernel to use updated packages.


In [10]:
from geopy.geocoders import Nominatim

# Initialize empty lists to store latitude and longitude values
latitudes = []
longitudes = []

# Iterate over each country name and get its coordinates
for country_name in unique_countries:
    geolocator = Nominatim(user_agent="geo_locator")
    location = geolocator.geocode(country_name)
    if location:
        latitudes.append(location.latitude)
        longitudes.append(location.longitude)
    else:
        # If coordinates not found, append None
        latitudes.append(None)
        longitudes.append(None)

# Print the latitude and longitude lists
print("Latitude values:", latitudes)
print("Longitude values:", longitudes)


Latitude values: [33.7680065, 1.2315264, 28.0000272, -14.297124, None, 42.5407167, -11.8775768, 17.2234721, -34.9964963, 4.536307, -30.0000769, -24.7761086, 47.59397, 40.3936294, 24.7736546, 26.1551249, -0.2864982, 13.1500331, 53.4250605, 50.6402809, 17.1204943, 9.5293472, 32.2937417, 27.549511, -17.0568696, 44.3053476, -23.1681782, -10.3333333, 4.4137155, 42.6073975, 12.0753083, -3.426449, 12.5433216, 4.6125522, 61.0666922, 16.0000552, 18.4, 7.0323598, 39.4009215, 35.6640736, None, None, None, 15.6134137, -31.7613365, 35.000074, 4.099917, -12.2045176, -2.9814344, 10.2735633, 7.9897371, 45.3658443, 23.0131338, 34.9174159, 49.7439047, -15.4139213, 55.670249, 11.8145966, 15.4113138, 19.0974031, 41.700789650000004, 50.26590325, None, -1.3397668, 26.2540493, 13.8000382, 52.5310214, 1.613172, 15.9500319, 58.7523778, 10.2116702, -18.1239696, 63.2467777, 46.603354, -0.8999695, 13.470062, 32.3293809, 51.1638175, 8.0300284, 43.2097838, 35.9948477, 12.1360374, 13.45011745, 15.5855545, 10.7226226

In [11]:
latitudes = [0 if x is None else x for x in latitudes]

longitudes = [0 if x is None else x for x in longitudes]

latitudes

longitudes


[66.2385139,
 -75.89204331320687,
 2.9999825,
 -170.7131481,
 0,
 1.5732033,
 17.5691241,
 -61.9554608,
 -64.9672817,
 -75.6723751,
 139.9998196,
 134.755,
 14.12456,
 47.7872508,
 -78.0000547,
 50.5344606,
 36.0514231,
 -59.5250305,
 27.6971358,
 4.6667145,
 -88.6859028,
 2.2584408,
 -64.7815286,
 90.5119273,
 -64.9912286,
 17.5961467,
 24.5928742,
 -53.2,
 114.5653908,
 25.4856617,
 -1.6880314,
 29.9324519,
 104.8144914,
 13.1535811,
 -107.991707,
 -24.0083947,
 -75.0,
 19.9981227,
 72.8676621,
 -95.1521353,
 0,
 0,
 0,
 19.0156172,
 -71.3187697,
 104.999927,
 -72.9088133,
 44.2832964,
 23.8222636,
 -84.0739102,
 -5.5679458,
 15.6575209,
 -80.8328748,
 32.889902651331866,
 15.3381061,
 28.289698,
 10.3333283,
 42.8453061,
 -61.3653618,
 -70.3028026,
 41.72010121238955,
 18.69757619439062,
 0,
 -79.3666965,
 29.2675469,
 -88.9140683,
 -1.2649062,
 10.5170357,
 37.9999668,
 25.3319078,
 38.6521203,
 179.0122737,
 25.9209164,
 1.8883335,
 11.6899699,
 -15.4900464,
 -83.1137366,
 10.4478

In [12]:
import requests
import pandas as pd


# API key (replace 'YOUR_API_KEY' with your actual API key)
api_key = '936c8c5a35fb206b715cdfcd87cba32b'

# Initialize an empty list to store data
data = []

# Iterate over latitudes and longitudes
for lat, lon in zip(latitudes, longitudes):
    # API endpoint URL
    url = f"http://api.openweathermap.org/data/2.5/air_pollution?lat={lat}&lon={lon}&appid={api_key}"
    
    # Send GET request
    response = requests.get(url)
    
    # Check if request was successful
    if response.status_code == 200:
        # Parse JSON response
        json_data = response.json()
        
        # Extract required data from JSON
        for item in json_data['list']:
            aqi = item['main']['aqi']
            dt = item['dt']
            co = item['components']['co']
            no = item['components']['no']
            no2 = item['components']['no2']
            o3 = item['components']['o3']
            so2 = item['components']['so2']
            pm2_5 = item['components']['pm2_5']
            pm10 = item['components']['pm10']
            nh3 = item['components']['nh3']
            
            # Append data to the list
            data.append({'lat': lat, 'lon': lon, 'aqi': aqi, 'dt': dt, 'co': co, 'no': no, 'no2': no2, 'o3': o3, 'so2': so2, 'pm2_5': pm2_5, 'pm10': pm10, 'nh3': nh3})

# Create DataFrame from the list of dictionaries
curr_data = pd.DataFrame(data)

# Print the DataFrame
curr_data


lat         lon  aqi          dt      co    no   no2     o3   so2  \
0    33.768006   66.238514    2  1713445979  188.59  0.00  0.06  75.10  0.01   
1     1.231526  -75.892043    1  1713445979  223.64  0.06  1.06   5.90  0.04   
2    28.000027    2.999983    3  1713445980  188.59  0.02  0.10  97.28  0.13   
3   -14.297124 -170.713148    1  1713445980  203.61  0.00  0.03  27.54  0.07   
4     0.000000    0.000000    1  1713445980  300.41  0.00  0.02  23.25  0.05   
..         ...         ...  ...         ...     ...   ...   ...    ...   ...   
226   0.000000    0.000000    1  1713445980  300.41  0.00  0.02  23.25  0.05   
227 -45.033025  168.661112    1  1713446052  195.27  0.00  0.51  35.76  0.16   
228  16.347124   47.891527    2  1713446052  186.92  0.01  0.12  33.98  0.08   
229 -14.518912   27.558988    1  1713446053  220.30  0.01  0.06  22.17  0.20   
230 -18.455496   29.746841    1  1713446053  216.96  0.03  0.16  31.11  0.63   

     pm2_5   pm10   nh3  
0     0.68   1.20  0.07  
1     1.52   1.66  0.38  
2    18.24  94.97  0.07  
3     2.74  11.61  0.30  
4     1.95   4.82  0.00  
..     ...    ...   ...  
226   1.95   4.82  0.00  
227   0.50   0.67  1.36  
228   5.97  25.73  0.18  
229   0.51   0.79  0.28  
230   1.29   2.31  0.18  

[231 rows x 12 columns]

In [13]:
curr_data['countries'] = unique_countries

curr_data


lat         lon  aqi          dt      co    no   no2     o3   so2  \
0    33.768006   66.238514    2  1713445979  188.59  0.00  0.06  75.10  0.01   
1     1.231526  -75.892043    1  1713445979  223.64  0.06  1.06   5.90  0.04   
2    28.000027    2.999983    3  1713445980  188.59  0.02  0.10  97.28  0.13   
3   -14.297124 -170.713148    1  1713445980  203.61  0.00  0.03  27.54  0.07   
4     0.000000    0.000000    1  1713445980  300.41  0.00  0.02  23.25  0.05   
..         ...         ...  ...         ...     ...   ...   ...    ...   ...   
226   0.000000    0.000000    1  1713445980  300.41  0.00  0.02  23.25  0.05   
227 -45.033025  168.661112    1  1713446052  195.27  0.00  0.51  35.76  0.16   
228  16.347124   47.891527    2  1713446052  186.92  0.01  0.12  33.98  0.08   
229 -14.518912   27.558988    1  1713446053  220.30  0.01  0.06  22.17  0.20   
230 -18.455496   29.746841    1  1713446053  216.96  0.03  0.16  31.11  0.63   

     pm2_5   pm10   nh3                   countries  
0     0.68   1.20  0.07                 Afghanistan  
1     1.52   1.66  0.38                     Albania  
2    18.24  94.97  0.07                     Algeria  
3     2.74  11.61  0.30              American Samoa  
4     1.95   4.82  0.00        Andean Latin America  
..     ...    ...   ...                         ...  
226   1.95   4.82  0.00  Western Sub-Saharan Africa  
227   0.50   0.67  1.36                       World  
228   5.97  25.73  0.18                       Yemen  
229   0.51   0.79  0.28                      Zambia  
230   1.29   2.31  0.18                    Zimbabwe  

[231 rows x 13 columns]

In [14]:
# Filtering out rows where both latitude and longitude are 0
curr_data = curr_data[(curr_data['lat'] != 0) & (curr_data['lon'] != 0)]


In [15]:
curr_data

lat         lon  aqi          dt      co    no   no2     o3   so2  \
0    33.768006   66.238514    2  1713445979  188.59  0.00  0.06  75.10  0.01   
1     1.231526  -75.892043    1  1713445979  223.64  0.06  1.06   5.90  0.04   
2    28.000027    2.999983    3  1713445980  188.59  0.02  0.10  97.28  0.13   
3   -14.297124 -170.713148    1  1713445980  203.61  0.00  0.03  27.54  0.07   
5    42.540717    1.573203    2  1713445981  185.25  0.06  0.23  85.83  0.11   
..         ...         ...  ...         ...     ...   ...   ...    ...   ...   
225  53.859829   27.553531    2  1713446051  223.64  0.12  1.65  78.68  0.98   
227 -45.033025  168.661112    1  1713446052  195.27  0.00  0.51  35.76  0.16   
228  16.347124   47.891527    2  1713446052  186.92  0.01  0.12  33.98  0.08   
229 -14.518912   27.558988    1  1713446053  220.30  0.01  0.06  22.17  0.20   
230 -18.455496   29.746841    1  1713446053  216.96  0.03  0.16  31.11  0.63   

     pm2_5   pm10   nh3       countries  
0     0.68   1.20  0.07     Afghanistan  
1     1.52   1.66  0.38         Albania  
2    18.24  94.97  0.07         Algeria  
3     2.74  11.61  0.30  American Samoa  
5     0.50   0.57  1.30         Andorra  
..     ...    ...   ...             ...  
225   0.50   0.57  0.98  Western Europe  
227   0.50   0.67  1.36           World  
228   5.97  25.73  0.18           Yemen  
229   0.51   0.79  0.28          Zambia  
230   1.29   2.31  0.18        Zimbabwe  

[216 rows x 13 columns]

In [16]:
curr_data = curr_data.reset_index(drop=True)

In [17]:
curr_data

lat         lon  aqi          dt      co    no   no2     o3   so2  \
0    33.768006   66.238514    2  1713445979  188.59  0.00  0.06  75.10  0.01   
1     1.231526  -75.892043    1  1713445979  223.64  0.06  1.06   5.90  0.04   
2    28.000027    2.999983    3  1713445980  188.59  0.02  0.10  97.28  0.13   
3   -14.297124 -170.713148    1  1713445980  203.61  0.00  0.03  27.54  0.07   
4    42.540717    1.573203    2  1713445981  185.25  0.06  0.23  85.83  0.11   
..         ...         ...  ...         ...     ...   ...   ...    ...   ...   
211  53.859829   27.553531    2  1713446051  223.64  0.12  1.65  78.68  0.98   
212 -45.033025  168.661112    1  1713446052  195.27  0.00  0.51  35.76  0.16   
213  16.347124   47.891527    2  1713446052  186.92  0.01  0.12  33.98  0.08   
214 -14.518912   27.558988    1  1713446053  220.30  0.01  0.06  22.17  0.20   
215 -18.455496   29.746841    1  1713446053  216.96  0.03  0.16  31.11  0.63   

     pm2_5   pm10   nh3       countries  
0     0.68   1.20  0.07     Afghanistan  
1     1.52   1.66  0.38         Albania  
2    18.24  94.97  0.07         Algeria  
3     2.74  11.61  0.30  American Samoa  
4     0.50   0.57  1.30         Andorra  
..     ...    ...   ...             ...  
211   0.50   0.57  0.98  Western Europe  
212   0.50   0.67  1.36           World  
213   5.97  25.73  0.18           Yemen  
214   0.51   0.79  0.28          Zambia  
215   1.29   2.31  0.18        Zimbabwe  

[216 rows x 13 columns]

In [30]:
import pandas as pd

# Assuming you have an existing DataFrame named 'original_df'
# Let's say it has columns: 'A', 'B', 'C', 'D', 'E'

# Create a new DataFrame with only columns 'A' and 'C'
selected_columns = ['lat', 'lon', 'countries']
forecast_data = curr_data[selected_columns]

# Now 'new_df' will only contain the columns 'A' and 'C' from 'original_df'
forecast_data

lat         lon       countries
0    33.768006   66.238514     Afghanistan
1     1.231526  -75.892043         Albania
2    28.000027    2.999983         Algeria
3   -14.297124 -170.713148  American Samoa
4    42.540717    1.573203         Andorra
..         ...         ...             ...
211  53.859829   27.553531  Western Europe
212 -45.033025  168.661112           World
213  16.347124   47.891527           Yemen
214 -14.518912   27.558988          Zambia
215 -18.455496   29.746841        Zimbabwe

[216 rows x 3 columns]

In [31]:
import requests
import pandas as pd

# Assuming your DataFrame is named df_lat_lon containing latitude and longitude pairs

# Function to make API request and calculate average values
def get_average_air_pollution(lat, lon, api_key):
    url = f"http://api.openweathermap.org/data/2.5/air_pollution/forecast?lat={lat}&lon={lon}&appid={api_key}"
    response = requests.get(url)
    data = response.json()
    
    # Initialize variables to store sum of values
    sum_co = sum_no = sum_no2 = sum_so2 = sum_o3 = sum_pm2_5 = sum_pm10 = sum_nh3 = sum_aqi = 0
    
    # Calculate sum of values
    for entry in data['list']:
        main = entry['main']
        components = entry['components']
        
        sum_co += components['co']
        sum_no += components['no']
        sum_no2 += components['no2']
        sum_so2 += components['so2']
        sum_o3 += components['o3']
        sum_pm2_5 += components['pm2_5']
        sum_pm10 += components['pm10']
        sum_nh3 += components['nh3']
        sum_aqi += main['aqi']
    
    # Calculate averages
    num_entries = len(data['list'])
    avg_co = sum_co / num_entries
    avg_no = sum_no / num_entries
    avg_no2 = sum_no2 / num_entries
    avg_so2 = sum_so2 / num_entries
    avg_o3 = sum_o3 / num_entries
    avg_pm2_5 = sum_pm2_5 / num_entries
    avg_pm10 = sum_pm10 / num_entries
    avg_nh3 = sum_nh3 / num_entries
    avg_aqi = sum_aqi / num_entries
    
    return avg_co, avg_no, avg_no2, avg_so2, avg_o3, avg_pm2_5, avg_pm10, avg_nh3, avg_aqi

# Function to process DataFrame and store average values
def process_dataframe(df, api_key):
    avg_values = []
    for index, row in df.iterrows():
        lat = row['lat']
        lon = row['lon']
        avg_values.append(get_average_air_pollution(lat, lon, api_key))
    avg_df = pd.DataFrame(avg_values, columns=['Average_CO', 'Average_NO', 'Average_NO2', 'Average_SO2', 'Average_O3', 'Average_PM2_5', 'Average_PM10', 'Average_NH3', 'Average_AQI'])
    return avg_df

# Call function to process DataFrame and store average values
api_key = "936c8c5a35fb206b715cdfcd87cba32b"
avg_forecast_data = process_dataframe(forecast_data, api_key)

# Display the resulting DataFrame
avg_forecast_data


Average_CO  Average_NO  Average_NO2  Average_SO2  Average_O3  \
0    220.785521    0.009583     0.236667     0.015833   59.351458   
1    233.406979    0.053437     0.751250     0.057083    5.580104   
2    197.768021    0.013438     0.218021     0.056875   80.206667   
3    202.462813    0.003438     0.039167     0.036979   20.392708   
4    185.233750    0.131667     1.199792     0.158229   79.714375   
..          ...         ...          ...          ...         ...   
211  247.035625    0.261146     5.674896     1.328646   74.986146   
212  199.837396    0.023750     0.799583     0.205417   30.704167   
213  200.672708    0.009688     0.142812     0.045625   29.644792   
214  219.882083    0.009479     0.188854     0.152708   19.549583   
215  227.495313    0.019896     0.429583     0.742292   30.307604   

     Average_PM2_5  Average_PM10  Average_NH3  Average_AQI  
0         1.035104      2.534167     0.225104     1.520833  
1         0.712604      0.801563     0.599271     1.000000  
2        47.201979    232.898646     0.240417     4.364583  
3         1.160938      3.814167     0.369583     1.000000  
4         2.870313      3.306979    12.102708     2.093750  
..             ...           ...          ...          ...  
211       2.193333      2.715000     2.429167     1.958333  
212       0.634896      1.336979     2.221875     1.000000  
213       2.263958      8.802708     0.087917     1.010417  
214       1.179167      1.715833     0.258125     1.000000  
215       3.058021      4.679687     0.408021     1.000000  

[216 rows x 9 columns]

In [32]:
combined_avg_forecast_data = pd.concat([avg_forecast_data, forecast_data], axis=1)
combined_avg_forecast_data


Average_CO  Average_NO  Average_NO2  Average_SO2  Average_O3  \
0    220.785521    0.009583     0.236667     0.015833   59.351458   
1    233.406979    0.053437     0.751250     0.057083    5.580104   
2    197.768021    0.013438     0.218021     0.056875   80.206667   
3    202.462813    0.003438     0.039167     0.036979   20.392708   
4    185.233750    0.131667     1.199792     0.158229   79.714375   
..          ...         ...          ...          ...         ...   
211  247.035625    0.261146     5.674896     1.328646   74.986146   
212  199.837396    0.023750     0.799583     0.205417   30.704167   
213  200.672708    0.009688     0.142812     0.045625   29.644792   
214  219.882083    0.009479     0.188854     0.152708   19.549583   
215  227.495313    0.019896     0.429583     0.742292   30.307604   

     Average_PM2_5  Average_PM10  Average_NH3  Average_AQI        lat  \
0         1.035104      2.534167     0.225104     1.520833  33.768006   
1         0.712604      0.801563     0.599271     1.000000   1.231526   
2        47.201979    232.898646     0.240417     4.364583  28.000027   
3         1.160938      3.814167     0.369583     1.000000 -14.297124   
4         2.870313      3.306979    12.102708     2.093750  42.540717   
..             ...           ...          ...          ...        ...   
211       2.193333      2.715000     2.429167     1.958333  53.859829   
212       0.634896      1.336979     2.221875     1.000000 -45.033025   
213       2.263958      8.802708     0.087917     1.010417  16.347124   
214       1.179167      1.715833     0.258125     1.000000 -14.518912   
215       3.058021      4.679687     0.408021     1.000000 -18.455496   

            lon       countries  
0     66.238514     Afghanistan  
1    -75.892043         Albania  
2      2.999983         Algeria  
3   -170.713148  American Samoa  
4      1.573203         Andorra  
..          ...             ...  
211   27.553531  Western Europe  
212  168.661112           World  
213   47.891527           Yemen  
214   27.558988          Zambia  
215   29.746841        Zimbabwe  

[216 rows x 12 columns]

In [20]:
gas_correlation = df[['co', 'no', 'no2', 'so2', 'Air pollution (total) (deaths per 100,000)']].corr()
print("\nHypothesis 4: Correlation between Gases (CO, NO, NO2, SO2) and Total Air Pollution-related deaths:")
gas_correlation


Hypothesis 4: Correlation between Gases (CO, NO, NO2, SO2) and Total Air Pollution-related deaths:


co        no       no2  \
co                                          1.000000  0.545838  0.605668   
no                                          0.545838  1.000000  0.688021   
no2                                         0.605668  0.688021  1.000000   
so2                                         0.303210  0.383231  0.436906   
Air pollution (total) (deaths per 100,000)  0.089807 -0.066360 -0.155235   

                                                 so2  \
co                                          0.303210   
no                                          0.383231   
no2                                         0.436906   
so2                                         1.000000   
Air pollution (total) (deaths per 100,000) -0.080222   

                                            Air pollution (total) (deaths per 100,000)  
co                                                                            0.089807  
no                                                                           -0.066360  
no2                                                                          -0.155235  
so2                                                                          -0.080222  
Air pollution (total) (deaths per 100,000)                                    1.000000

### There is a positive correlation between CO (Carbon Monoxide) and Air pollution-related deaths, with a correlation coefficient of approximately 0.09. This suggests a weak positive relationship between CO levels and the number of deaths caused by air pollution.
### There is a positive correlation between NO (Nitric Oxide) and Air pollution-related deaths, with a correlation coefficient of approximately 0.55. This indicates a moderate positive relationship between NO levels and the number of deaths caused by air pollution.
### There is a positive correlation between NO2 (Nitrogen Dioxide) and Air pollution-related deaths, with a correlation coefficient of approximately 0.61. This suggests a moderate positive relationship between NO2 levels and the number of deaths caused by air pollution.
### There is a weak positive correlation between SO2 (Sulfur Dioxide) and Air pollution-related deaths, with a correlation coefficient of approximately 0.08. This suggests a very weak positive relationship between SO2 levels and the number of deaths caused by air pollution.
### Considering these correlations, it seems that there is some support for this hypothesis, indicating that higher levels of certain gases like CO, NO, and NO2 are associated with increased deaths caused by air pollution. However, the strength of the correlation varies for each gas, with NO and NO2 showing stronger positive correlations compared to CO and SO2.

In [23]:
pm_correlation = df[['pm2_5', 'pm10', 'Air pollution (total) (deaths per 100,000)']].corr()
print("\nHypothesis 3: Correlation between PM2.5, PM10, and Total Air Pollution-related deaths:")
pm_correlation


Hypothesis 3: Correlation between PM2.5, PM10, and Total Air Pollution-related deaths:


pm2_5      pm10  \
pm2_5                                       1.000000  0.832156   
pm10                                        0.832156  1.000000   
Air pollution (total) (deaths per 100,000)  0.273780  0.247576   

                                            Air pollution (total) (deaths per 100,000)  
pm2_5                                                                         0.273780  
pm10                                                                          0.247576  
Air pollution (total) (deaths per 100,000)                                    1.000000

### There is a strong positive correlation between PM2.5 (Particulate Matter with a diameter of 2.5 micrometers or less) and Air pollution-related deaths, with a correlation coefficient of approximately 0.274. This suggests a moderate positive relationship between PM2.5 levels and the number of deaths caused by air pollution.
### There is also a strong positive correlation between PM10 (Particulate Matter with a diameter of 10 micrometers or less) and Air pollution-related deaths, with a correlation coefficient of approximately 0.248. This indicates a moderate positive relationship between PM10 levels and the number of deaths caused by air pollution.
### Considering these correlations, it appears that there is support for this hypothesis . Higher levels of particulate matter, both PM2.5 and PM10, are associated with increased deaths caused by air pollution. The strong positive correlations suggest that elevated levels of particulate matter in the air can contribute to a higher mortality rate due to air pollution.

In [36]:
df.dtypes

countries                                           object
Code                                                object
Year                                                 int64
Air pollution (total) (deaths per 100,000)         float64
Indoor air pollution (deaths per 100,000)          float64
Outdoor particulate matter (deaths per 100,000)    float64
Outdoor ozone pollution (deaths per 100,000)       float64
starttime                                            int64
endtime                                              int64
Latitude                                           float64
Longitude                                          float64
co                                                 float64
no                                                 float64
no2                                                float64
o3                                                 float64
so2                                                float64
pm2_5                                              float

In [37]:
curr_data.dtypes

lat          float64
lon          float64
aqi            int64
dt             int64
co           float64
no           float64
no2          float64
o3           float64
so2          float64
pm2_5        float64
pm10         float64
nh3          float64
countries     object
dtype: object

In [38]:
combined_avg_forecast_data.dtypes

Average_CO       float64
Average_NO       float64
Average_NO2      float64
Average_SO2      float64
Average_O3       float64
Average_PM2_5    float64
Average_PM10     float64
Average_NH3      float64
Average_AQI      float64
lat              float64
lon              float64
countries         object
dtype: object

In [33]:
import pandas as pd
import sqlite3


# SQLite database file
db_file = 'data.db'

# Connect to the database
conn = sqlite3.connect(db_file)

# Store DataFrames in the SQLite database
df.to_sql('Historical Data Table', conn, index=False, if_exists='replace')
curr_data.to_sql('Current Data Table', conn, index=False, if_exists='replace')
combined_avg_forecast_data.to_sql('Forecast Data Table', conn, index=False, if_exists='replace')

# Close the connection
conn.close()

print("DataFrames stored in SQLite database successfully.")


DataFrames stored in SQLite database successfully.
